lib/src/infinite_list.dart

```diff
@@ -0,0 +1,253 @@
+  /// {@endtemplate}
+  final bool isLoading;
+
+  /// {@template has_error}
+  /// Indicates if an error has occurred.
+  ///
+  /// While set to `true`, the [onFetchData] callback will not be triggered
+  /// and the [errorBuilder] will be rendered.
+  ///
+  /// Is set to `false` by default and cannot be `null`.
+  /// {@endtemplate}
+  final bool hasError;
+
+  /// {@template has_reached_max}
+  /// Indicates if the end of the data source has been reached and no more
+  /// data can be fetched.
+  ///
+  /// While set to `true`, the [onFetchData] callback will not be triggered.
+  ///
+  /// Is set to `false` by default and cannot be `null`.
+  /// {@endtemplate}
+  final bool hasReachedMax;
+
+  /// {@template on_fetch_data}
+  /// The callback method that's called whenever the list is scrolled to the end
+  /// (meaning the top when [reverse] is `true`, or the bottom otherwise).
+  ///
+  /// In normal operation, this method should trigger new data to be fetched and
+  /// [isLoading] to be set to `true`.
+  ///
+  /// Exactly when this is called depends on the [scrollExtentThreshold].
+  /// Additionally, every call to this will be debounced by the provided
+  /// [debounceDuration].
+  ///
+  /// Is required and cannot be `null`.
+  /// {@endtemplate}
+  final VoidCallback onFetchData;
+
+  /// {@template padding}
+  /// The amount of space by which to inset the list of items.
+  ///
+  /// Is optional and can be `null`.
+  /// {@endtemplate}
+  final EdgeInsets? padding;
+
+  /// {@template empty_builder}
+  /// An optional builder that's shown when the list of items is empty.
+  ///
+  /// If `null`, nothing is shown.
+  /// {@endtemplate}
+  final WidgetBuilder? emptyBuilder;
+
+  /// {@template loading_builder}
+  /// An optional builder that's shown at the end of the list when [isLoading]
+  /// is `true`.
+  ///
+  /// Defaults to [defaultInfiniteListLoadingBuilder].
+  /// {@endtemplate}
+  final WidgetBuilder? loadingBuilder;
+
+  /// {@template error_builder}
+  /// An optional builder that's shown when [hasError] is not `null`.
+  ///
+  /// Defaults to [defaultInfiniteListErrorBuilder].
+  final WidgetBuilder? errorBuilder;
+
+  /// {@template separator_builder}
+  /// An optional builder that, when provided, is used to show a widget in
+  /// between every pair of items.
+  ///
+  /// If the [itemBuilder] returns a [ListTile], this is commonly used to render
+  /// a [Divider] between every tile.
+  ///
+  /// Is optional and can be `null`.
+  /// {@endtemplate}
+  final WidgetBuilder? separatorBuilder;
+
+  /// {@template item_builder}
+  /// The builder used to build a widget for every index of the `itemCount`.
+  ///
+  /// Is required and cannot be `null`.
+  /// {@endtemplate}
+  final ItemBuilder itemBuilder;
+
+  // To work just as a plain ListView, It should automatically apply a
+  // media query padding if the passing options is omitted or null.
+  Widget _buildSliver(BuildContext context) {
```

Comment:
Can we extract this to a `StatelessWidget`?

```dart
class _Sliver extends StatelessWidget {
  @override
  Widget build(BuildContext context) {...}
}
```

Reply:
Great idea. I've encasulated padding into its own widget. _ContextualSliverPadding.